### ISSUE TITLE
Use tls-crypt instead of tls-auth

### ISSUE THREAD

**Issue by francoism90** [graysky2/ovpngen#4]
On the Arch Wiki `tls-crypt` is used as example, you'll need to change the `<tls-auth>` to `<tls-crypt>` to get it to work.

Would it be possible to set/implement this into ovpngen?

Thanks.

**Comment by graysky2** [comment]
When I take my functional openvpn server.conf and make this substitution:
```
#tls-auth /etc/openvpn/server/ta.key 0 # This file is secret
tls-crypt /etc/openvpn/server/ta.key 0 # This file is secret
```

I receive Transport Errors from the client using an unmodified profile.  If I edit the profile changing this, I get the same errors and no connection.  Perhaps I do not understand how to correctly enable the `tls-crypt` option?

```
</key>
<tls-crypt>
#
# 2048 bit OpenVPN static key
#
-----END OpenVPN Static key V1-----
</tls-crypt>
...
```

**Comment by francoism90** [comment]
Strange, I don't have any issues, only when using `tls-auth` together with `tls-crypt`.
```
port 1194
proto udp
dev tun
ca ca.crt
cert server.crt
key server.key
dh dh.pem
tls-crypt ta.key
server 10.8.0.0 255.255.255.0
ifconfig-pool-persist ipp.txt
keepalive 10 120
cipher AES-256-CBC
auth SHA512
compress lz4-v2
push "compress lz4-v2"
user nobody
group nobody
persist-key
persist-tun
status openvpn-status.log
verb 3
mute 20
explicit-exit-notify 1
```

client.ovpn
```
client
dev tun
remote example.org 1194 udp
resolv-retry infinite
nobind
;user nobody
;group nobody
persist-key
persist-tun
verb 3
cipher AES-256-CBC
auth SHA512
compress lz4
remote-cert-tls server
key-direction 1
<ca>
-----BEGIN CERTIFICATE-----
-----END CERTIFICATE-----
</ca>
<cert>
-----BEGIN CERTIFICATE-----
-----END CERTIFICATE-----
</cert>
<key>
-----BEGIN PRIVATE KEY-----
-----END PRIVATE KEY-----
</key>
<tls-crypt>
#
# 2048 bit OpenVPN static key
#
-----BEGIN OpenVPN Static key V1-----

-----END OpenVPN Static key V1-----
</tls-crypt>
```

What kind of errors do you get?

**Comment by graysky2** [comment]
Grrr... my bad.  Version 1.2.5 of the iOS app has a new behavior (bug?): when users switch profiles it deems the new one "not enabled" and defaults back to the last-used profile.  One has to go through extra steps to enable them.

I can now verify the behavior.  The catch is that unlike encryption, the server and client profiles must match in their usage of `tls-auth` or `tls-crypt` in my testing.  I am not sure how to make this script smart to see that without forcing the user to add another token corresponding to the `server.conf` and use some grep/sed logic to pick it.  I suppose I can create a comment instructing the user to manually edit the created `foo.ovpn` file after it is successfully generated.

EDIT: When Look into `/usr/share/openvpn/examples/server.conf` provided by openvpn 2.4.4-1, I see no mention of `tls-crypt` therein.  My though is to leave the script as-is just adding the comment.

Thoughts?

**Comment by francoism90** [comment]
@graysky2 Good solution! If I'm not mistaken the official Android OpenVPN app doesn't support `tls-crypt` at the moment, although it works fine in the _OpenVPN for android_.

The documentation and examples should be improved. It also doesn't mention the new compression options. Maybe it will be better in the next release, on the roadmap it seems they want to do a clean-up.

Thanks for the testing. :)

PS. Maybe we should change the (default) `tls-crypt` in the ArchWiki back to `tls-auth` for now.

**Comment by graysky2** [comment]
No idea about the 'droid version.  I wouldn't edit the AW article removing the option since the Linux and iOS versions support it (droid support unverified).  Just add a note about that.

As I stated above, the compression options are server-side implemented; the client profile doesn't need to have anything listed.

EDIT: for illustrative purposes, my server config:
```
cipher AES-256-CBC
auth SHA512
tls-version-min 1.2
tls-cipher TLS-DHE-RSA-WITH-AES-256-GCM-SHA384:TLS-DHE-RSA-WITH-AES-128-GCM-SHA256:TLS-DHE-RSA-WITH-AES-256-CBC-SHA:TLS-DHE-RSA-WITH-CAMELLIA-256-CBC-SHA:TLS-DHE-RSA-WITH-AES-128-CBC-SHA:TLS-DHE-RSA-WITH-CAMELLIA-128-CBC-SHA
port 443
proto tcp
dev tun
ca /etc/openvpn/server/ca.crt
cert /etc/openvpn/server/server.crt
key /etc/openvpn/server/server.key
dh /etc/openvpn/server/dh.pem
topology subnet
server 10.8.0.0 255.255.255.0
ifconfig-pool-persist ipp.txt
# next 2 lines for for pihole
push "redirect-gateway def1 bypass-dhcp"
push "dhcp-option DNS 10.1.1.250"
keepalive 10 120
tls-crypt /etc/openvpn/server/ta.key 0
compress lz4-v2
push "compress lz4-v2"
max-clients 4
user nobody
group nobody
persist-key
persist-tun
status openvpn-status.log
verb 3
```

And my client config:
```
client
dev tun
remote my.cool.domain.com 443 tcp
resolv-retry infinite
nobind
persist-key
persist-tun
verb 3
cipher AES-256-CBC
auth SHA512
remote-cert-tls server
key-direction 1
<ca>
 ...
</ca>
<cert>
 ...
</cert>
<key>
 ...
</key>
<tls-crypt>
 ...
</tls-crypt>
```

**Comment by slokhorst** [comment]
ArchWiki editor here. I also ran into the problem that I couldn't connect with the Android app.

Since you seem to know more about the subject, could you maybe comment [here](https://wiki.archlinux.org/index.php/Talk:OpenVPN#tls-crypt_as_default:_is_it_worth_it.3F)?

**Comment by graysky2** [comment]
Done... not much help though :/

**Comment by slokhorst** [comment]
Thanks both! Also thanks @graysky2 for this very useful tool. :)

**Comment by francoism90** [comment]
@slokhorst Did you test the open-source Android app instead of the official one?

**Comment by slokhorst** [comment]
Ugh I'm so sorry! I was confused by these reports on the web, but it turns out that tls-crypt **is supported** in the official OpenVPN Connect Android app [since earlier this year](https://github.com/OpenVPN/openvpn3/issues/11#issuecomment-362891509). 

I had messed up my `.ovpn` file when changing `<tls-auth>` to `<tls-crypt>`; that's why it didn't work. I'll cleanup the stuff I added to the wiki article and try to clarify things so other people don't make the same mistake I made.

To get back to the original report: maybe it's an idea to add an optional `--tls-crypt` flag to `ovpngen` that will make the substitution. And if it's not specified it just keeps on using tls-auth?

Again sorry all for the confusion.
